### PR TITLE
Support PHPUnit 10

### PR DESCRIPTION
**Pull request by spawnia**
The changes in `ArraySubsetAsserts` were made due to the following test failures:

```
There were 2 failures:

1) DMS\PHPUnitExtensions\ArraySubset\Tests\Unit\AssertTest::testAssertArraySubsetThrowsExceptionForInvalidSubsetArgument
Failed asserting that exception of type "Error" matches expected exception "PHPUnit\Framework\Exception". Message was: "Call to undefined method PHPUnit\Framework\InvalidArgumentException::create()" at
/home/bfranke/projects/phpunit-arraysubset-asserts/src/ArraySubsetAsserts.php:35
/home/bfranke/projects/phpunit-arraysubset-asserts/tests/Unit/AssertTest.php:32
.

2) DMS\PHPUnitExtensions\ArraySubset\Tests\Unit\AssertTest::testAssertArraySubsetThrowsExceptionForInvalidArrayArgument
Failed asserting that exception of type "Error" matches expected exception "PHPUnit\Framework\Exception". Message was: "Call to undefined method PHPUnit\Framework\InvalidArgumentException::create()" at
/home/bfranke/projects/phpunit-arraysubset-asserts/src/ArraySubsetAsserts.php:51
/home/bfranke/projects/phpunit-arraysubset-asserts/tests/Unit/AssertTest.php:38
```

### CODE REVIEW

**Comment by jrfnl** [comment]
Looks like you are removing bits and pieces which ensured compatibility with PHPUnit 8.x. I very much doubt CI will pass on this PR.

**Comment by spawnia** [comment]
> Looks like you are removing bits and pieces which ensured compatibility with PHPUnit 8.x. I very much doubt CI will pass on this PR.

Indeed, there are behavioural changes. I believe there is little value in maintaining the different "original" error messages or exception types (if there even is such a thing, given this feature has been missing from PHPUnit for quite some time). It is not like someone is going to build functionality around what specifically happens when `assertArraySubset()` is misused in a certain way.

I don't see any incompatibility being introduced. The functionality can be used just like before, and the new error message should be just as clear as the one that was previously given.

Tests passed locally.

**Comment by jrfnl** [comment]
> Tests passed locally.

@spawnia Against which PHP/PHPUnit combinations ?

**Comment by mattvb91** [comment]
@spawnia can you trigger the tests in your repo?

**Comment by spawnia** [comment]
I was able to run the tests in my fork and fixed a few minor issues that prevented CI from working, see https://github.com/spawnia/phpunit-arraysubset-asserts/pull/1.

However, there is still a CI failure due to usage of a deprecated configuration schema, see https://github.com/spawnia/phpunit-arraysubset-asserts/actions/runs/4112652928/jobs/7097833112. I tried to fix this with [`b5299be` (#70)](https://github.com/rdohms/phpunit-arraysubset-asserts/pull/70/commits/b5299be629a63cd3013c2eac02b9bb6278271a3d), but apparently this did not help. I was able to fix the issue by basically miniziming the PHPUnit configuration to avoid usage of incompatible elements (e.g. `whitelist` was changed to `include`): [`97c61a5` (#70)](https://github.com/rdohms/phpunit-arraysubset-asserts/pull/70/commits/97c61a56be6043adba7ad6451e0fd9b3824cb7b5).

**Comment by mattvb91** [comment]
@jrfnl @rdohms any idea what is needed to get this in? Thanks

**Comment by jrfnl** [comment]
> @jrfnl @rdohms any idea what is needed to get this in? Thanks

Time for the maintainers to do a proper review. I'll do one once I'm back from my break, but that will still be a couple more weeks.

File: phpunit.xml.dist
Comment (jrfnl) [inline phpunit.xml.dist:15]: It's unclear to me why the majority of these settings are being removed.

AFAICS, the only change needed is: `xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.2/phpunit.xsd"` to indicate to PHPUnit the schema used (with 9.2 deliberately chosen as PHPUnit changes some thing in PHPUnit 9.3).

The original schema setting made it dependent on the Composer installed PHPUnit version, which makes the schema setting unstable if PHPUnit 10.0 would also be supported (as PHPUnit 10 made significant changes to the schema).

As for the other values - in part - being set to the default value used by PHPUnit: those defaults have changed for a lot of these across PHPUnit versions, which is why it is important to have these settings explicitly set to make sure tests are being run with the same configuration across PHPUnit versions.

File: src/ArraySubsetAsserts.php
Comment (jrfnl) [inline src/ArraySubsetAsserts.php:9]: These conditions were introduced in PR #65 in response to PR #50 and have a function (prevent the PHPUnit deprecation warning being thrown when using PHPUnit 8.x),

Using the generic PHP native `InvalidArgumentException` might break (external) tests using the polyfill, which rely on the PHPUnit native behaviour (which the current code closely emulates).

I understand changes may be needed to this logic toggle to allow for PHPUnit 10.0, but in that case, IMO, that should be handled in an additional toggle, not by breaking the existing functionality.

File: tests/Unit/AssertTest.php
Comment (jrfnl) [inline tests/Unit/AssertTest.php:7]: Dito, same as above.

**Review by jrfnl — COMMENTED** [review]
FWIW: I've done a detailed review of this PR.

In my personal opinion, this PR should not be merged as-is. Please see the inline comments for more detailed feedback.

File: src/ArraySubsetAsserts.php
Comment (jrfnl) [reply src/ArraySubsetAsserts.php:9]: Thinking this over some more.... I can imagine that a choice could be made to no longer use the PHPUnit native InvalidArgumentException/Helper.

In that case, I would recommend:
1. For this change to be annotated as a breaking change in the changelog, albeit one which will affect very few users.
2. To introduce an `InvalidArgumentException` class native to this package to handle the creation of the message.

File: .github/workflows/run-tests.yml
Comment (spawnia) [reply .github/workflows/run-tests.yml:27]: https://github.com/rdohms/phpunit-arraysubset-asserts/pull/73

File: phpunit.xml.dist
Comment (spawnia) [reply phpunit.xml.dist:15]: I reduced the PHPUnit configuration to be as minimal as possible in order to maximize compatbility between the different supported versions. Fortunately, the schema is not required and can simply be omitted.

File: src/ArraySubsetAsserts.php
Comment (spawnia) [reply src/ArraySubsetAsserts.php:9]: 1. Mentioning this in the changelog sounds good - I would add it, but found that you do not maintain it as a document?
2. So add `class InvalidArgumentException extends \InvalidArgumentException`? I can make that change, but honestly see no benefit. Why not keep things as simple as possible?

File: composer.json
Comment (jrfnl) [reply composer.json:7]: I personally prefer verbose as it reduces support issues, but 🤷🏻‍♀️ .

File: phpunit.xml.dist
Comment (spawnia) [reply phpunit.xml.dist]: Do you have a suggestion on how to fix https://github.com/spawnia/phpunit-arraysubset-asserts/actions/runs/4466219397/jobs/7844126983 then?

Comment (jrfnl) [reply phpunit.xml.dist]: Of course 😉 

I believe this can be fixed by running `phpunit --migrate-configuration` in a separate CI step prior to running the actual tests.
The step should be run conditionally for PHPUnit 10 only.

You can grab the PHPUnit version with something like this:
```yaml
      - name: Grab PHPUnit version
        id: phpunit_version
        run: echo "VERSION=$(vendor/bin/phpunit --version | grep --only-matching --max-count=1 --extended-regexp '\b[0-9]+\.[0-9]+')" >> $GITHUB_OUTPUT
```

... which allows you to check the version in a conditional step after that like so:
```yaml
        if: ${{ steps.phpunit_version.outputs.VERSION >= '10.0' }}
```

As a side-note: I find it weird that PHPUnit would exit with a non-zero exit code when it is only a warning from PHPUnit 10 itself and there are no test failures.
Might be worth opening an issue in the PHPUnit repo about this if there isn't one open already) as this is undesired and unexpected behaviour.

Comment (spawnia) [reply phpunit.xml.dist]: I could not get your suggestion to work, but found an alternative.

Comment (jrfnl) [reply phpunit.xml.dist]: I've been doing some testing with the `--migrate-configuration` option and have found that unfortunately it does not sufficiently convert the configuration.

Most notably, the PHPUnit 10.x `displayDetailsOn*` settings do not get set, while for a tool like this, which is used in the test pipeline by other projects, it is important to have access to information about deprecations in new PHP versions early.

For the PHPUnit Polyfills, I've now solved this by having a separate config file for PHPUnit 10.
If you'd like to see the implementation as an example, please have a look at: https://github.com/Yoast/PHPUnit-Polyfills/pull/111

Comment (jrfnl) [reply phpunit.xml.dist]: Just noticed that in the upcoming PHPUnit 10.1.0 release there will be new `failOnWarning`, `failOnNotice` and `failOnDeprecation` attributes. Those seem even better/more appropriate.

See:
* sebastianbergmann/phpunit#5196
* https://github.com/sebastianbergmann/phpunit/commit/fb6673f06ef90b43667b7187722e2c840dcdb573

**Comment by hilioski** [comment]
Any updates on this PR?

File: src/ArraySubsetAsserts.php
Comment (rdohms) [reply src/ArraySubsetAsserts.php:9]: Is the issue here that PHPUnit 10 removes the `PHPUnit\Util\InvalidArgumentHelper` class?
I believe in that case that we should indeed throw a generic `\InvalidArgumentException` but only if the Helper is not present.

File: composer.json
Comment (rdohms) [reply composer.json:7]: Let's be verbose, indeed.

**Review by rdohms — COMMENTED** [review]
Let's cater to @jrfnl comments and then clean up this branch. It has a lot of back-and-forth commits.
I'm happy to do that since I took so long to review it. I need some help understanding the Exception issue.

File: phpunit.xml.dist
Comment (rdohms) [reply phpunit.xml.dist]: We should tackle 10.1 in another PR and continue this discussion there.
WDYT?

**Comment by rdohms** [comment]
Alright, I did some cleanup and adjusted the Exception checks.
It should be all good now; @jrfnl can you re-review to see if I missed anything?

**Review by rdohms — APPROVED** [review]

**Comment by rdohms** [comment]
I also dropped the `composer.lock` once and for all.

**Review by spawnia — COMMENTED** [review]
Thanks @rdohms for getting into this, I am fine with the changes you made and hope it can get merged soon.

**Comment by dramamask** [comment]
I hope @jrfnl can review this soon. I'd love to keep using this package.

**Comment by slaff** [comment]
@jrfnl any chance of reviewing and merging this PR?

**Comment by GityaMan** [comment]
@jrfnl Is this package supported?

**Comment by alexw23** [comment]
Didn't see this getting merged soon - using the help of Github Copilot was able to rewrite our 4 tests and remove an additional package.

**Comment by senaranya** [comment]
> Didn't see this getting merged soon - using the help of Github Copilot was able to rewrite our 4 lines using this and remove an additional package.

Guess I may have to do the same! Can you please share a gist on how you did that?

**Comment by spawnia** [comment]
Thanks for the merge, waiting for the release now.

**Pull request merged**
